plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/util/HudiAvroSerializer.java

```diff
@@ -235,7 +249,57 @@ else if (javaType == Slice.class) {
                 writeSlice(output, type, value);
             }
             else if (javaType == LongTimestamp.class) {
-                type.writeObject(output, value);
+                if (value instanceof SqlTimestamp sqlTimestamp) {
+                    // value is read from parquet
+                    // From tests, sqlTimestamp is a UTC epoch that is converted from ZoneId#systemDefault()
+                    // IMPORTANT: Even when session's zoneId != ZoneId#systemDefault(), ZoneId#systemDefault() is used calculate/produce the false UTC.
+                    // The current sqlTimestamp is calculated as such:
+                    // 1. The true UTC timestamp that is stored in file is assumed to be in the local timezone
+                    // 2. Trino will them attempt to convert this to a false UTC by subtracting the timezone's offset (factoring offset rules like DST)
+                    // Hence, to calculate the true UTC, we will just have to reverse the steps
+
+                    // Reconstruct the original local wall time from sqlTimestamp's fields
+                    long microsFromSqlTs = sqlTimestamp.getEpochMicros();
+                    // picosFromSqlTs is defined as "picoseconds within the microsecond" (0 to 999,999)
+                    int picosFromSqlTs = sqlTimestamp.getPicosOfMicros();
+                    long secondsComponent = microsFromSqlTs / 1_000_000L;
+                    // Storing nanos component separately from seconds component, hence the modulo to remove secondsComponent
+                    int nanosComponent = (int) ((microsFromSqlTs % 1_000_000L) * 1000L + picosFromSqlTs / 1000L);
+                    LocalDateTime originalLocalWallTime = LocalDateTime.ofEpochSecond(secondsComponent, nanosComponent, ZoneOffset.UTC);
+
+                    // Determine the ZoneId in which originalLocalWallTime was observed
+                    ZoneId assumedOriginalZoneId = ZoneId.systemDefault();
+
+                    // Convert to true UTC by interpreting the local wall time in its original zone
+                    // This correctly handles DST for that zone at that specific historical date/time.
+                    ZonedDateTime zdtInOriginalZone;
```

Comment:
Do you have a test case on a super large seconds and picos in a timestamp to guard that there is no overflow in the calculation?

Reply:
Got it, looks like a minor issue for now.